proposals/NNNN-swiftpm-version-pinning.md

```diff
@@ -0,0 +1,277 @@
+
+ The `--reason` option is an optional argument to document the reason for pinning a dependency. This could be helpful for user to later remember why a dependency was pinned. Example:   
+ 
+	`$ swift package pin Foo --reason "The patch updates for Foo are really unstable and need screening."`
+
+
+2. Dependencies are never automatically pinned, pinning is only ever taken as a result of an explicit user action.
+ 
+3. We will add a new command `unpin`:
+
+	```
+	$ swift package unpin ( [--all] | [<package-name>] )
+	``` 
+	This is the counterpart to the pin command, and unpins one or all packages.
+
+4. We will fetch and resolve the dependencies when running the pin commands, in case we don't have the complete dependency graph yet.
+
+5. We will extend the workflow for update to honour version pinning, that is, it will only update packages which are unpinned, and it will only update to versions which can satisfy the existing pins. The update command will, however, also take an optional argument `--repin`:
+
+	```
+	$ swift package update [--repin]
+	```
+
+	* Update command errors if there are no unpinned packages which can be updated.
+
+	* Otherwise, the behaviour is to update all unpinned packages to the latest possible versions which can be resolved while respecting the existing pins.
+
+	* The `[--repin]` argument can be used to lift the version pinning restrictions. In this case, the behaviour is that all packages are updated, and packages which were previously pinned are then repinned to the latest resolved versions.
+
+6. The update and checkout will both emit logs, notifying the user that pinning is in effect.
+
+7. The `swift package show-dependencies` subcommand will be updated to indicate if a dependency is pinned.
+
+## Future Directions
+
+We have intentionally kept the pin file format an implementation detail in order
+to allow for future expansion. For example, we would like to consider embedding
+additional information on a known tag (like its SHA, when using Git) in the pins
+file as a security feature, to prevent man-in-the-middle attacks on parts of the
+package graph.
+
+## Impact on existing code
+
+There will be change in the behaviours of `swift build` and `swift package update` in presence of the pins file, as noted in the proposal however the existing package will continue to build without any modifications.
+
+## Alternative considered
+
+### Pin by default
+
+Much discussion has revolved around a single policy-default
+question: whether SwiftPM should generate a pins file as a matter
+of course any time it builds. This is how some other package
+managers work, and it is viewed as a conservative stance
+with respect to making repeatable builds more likely between
+developers. Developers will see the pins file and will be likely
+to check it in to their SCM system as a matter of convention.
+
+While pinning does reduce the risk of packages failing to build,
```

Comment:
Since our pinning design doesn't respect the pins of packages you depend on, I don't see how pinning can cause overconstraint dependency hell situations for us.